### ISSUE TITLE
Workflow with docker-compose in pycharm

### ISSUE THREAD

**Issue by Euphorbium** [tiangolo/fastapi#146]
How should I set up the run command so the debugger works, when the interpreter is in a docker image?

**Comment by euri10** [comment]
you have loads of possibilities depending of what you want to achieve, but overall you need for all of those to have your python interpreter set to the python inside docker, here's an example, you setup that in `Settings > Project > Project Interpreter > + > Docker-Compose`

![Imgur](https://i.imgur.com/khIa7Hm.png)

Then you may need to setup the path mappings if breakpoint don't get hit

![Imgur](https://i.imgur.com/NG968sA.png)

then you may have for instance a `manage.py` that runs uvicorn that way (reload is important):

```
@cli.command()
def rundev():
    uvicorn.run("app.app:app", host="0.0.0.0", logger=logging.getLogger("uvicorn"), reload=True)
```
and you hit the debug button running the rundev command of `manage.py` script :

![Imgur](https://i.imgur.com/dlSMhbS.png)

**Comment by Euphorbium** [comment]
So the main trick is setting up the manage.py file. Maybe this should be specified in the documentation, and a minimal manage.py file could be included.

**Comment by Euphorbium** [comment]
I am able to run django in docker compose, but not successful running  fastapi. I suppose some path mappings might be missing, but none were necessary for django. I am getting on run or debug.
```
Aborting on container exit...

Process finished with exit code 0
```

**Comment by euri10** [comment]
> So the main trick is setting up the manage.py file. Maybe this should be specified in the documentation, and a minimal manage.py file could be included.

not at all, this is just one of the endless possibilities, it's just an example, you may run it trough you `__main__` function, you may run it through another script, your may run it through `uvicorn` directly...as I said the most important is getting your interpreter right as described above


> I am able to run django in docker compose, but not successful running fastapi. I suppose some path mappings might be missing, but none were necessary for django. I am getting on run or debug.
> 
> ```
> Aborting on container exit...
> 
> Process finished with exit code 0
> ```

maybe a more detailed explanation of what you're doing would be a good the first step in getting help, not clear at all with the info provided

**Comment by tiangolo** [comment]
Thanks for your help @euri10 !

@Euphorbium for a ready-made project generator with Docker integrated, check: https://github.com/tiangolo/full-stack-fastapi-postgresql/

How are you connecting your debugger to Django inside of the container?

**Comment by TheGreatRefrigerator** [comment]
As i'm currently not using jango, the part for defining a cli command didn't work:
```
@cli.command()
def rundev():
    uvicorn.run("app.app:app", host="0.0.0.0", logger=logging.getLogger("uvicorn"), reload=True)
```

Debugging did also not work with starting the uvicorn server from the docker-compose.yml
`entrypoint: "start-reload.sh"`
or something like
`command: ["uvicorn", "main:app", "--host", "0.0.0.0", "--debug", "--port", "80", "--reload-dir", "/app"]`
([source](https://williamhayes.medium.com/fastapi-starlette-debug-vs-prod-5f7561db3a59))

For me debug was only working when doing it as [explained above](https://github.com/tiangolo/fastapi/issues/146#issuecomment-481582104) by euri10, but in `manage.py`:
```
import uvicorn

if __name__ == '__main__':
    uvicorn.run("app.main:app",
                host="0.0.0.0",
                port=80,
                log_level="debug",
                reload_dirs=["/app"])
```
Then after you started your docker, run `manage.py` in debug mode.

**Comment by jbagrible** [comment]
@TheGreatRefrigerator and @Euphorbium 
This really helped me get going. Thanks a lot.
